File: tools/cmpr_smoke_test.py
#!/usr/bin/env python3
"""
cmpr_smoke_test.py — independent Python verification of the CMPR encoder

Independently re-implements the algorithm from wii_cmpr.c in Python and runs
the same test cases.  Because it is a fresh implementation it will catch
logical errors that a C-only test might miss.

Run:
    python3 cmpr_smoke_test.py
"""

import struct
import sys

# ---------------------------------------------------------------------------
# Python re-implementation (mirrors wii_cmpr.c exactly)
# ---------------------------------------------------------------------------

def pack_rgb565(r: int, g: int, b: int) -> int:
    """Round-to-nearest RGB888 → RGB565."""
    r5 = (r * 31 + 127) // 255
    g6 = (g * 63 + 127) // 255
    b5 = (b * 31 + 127) // 255
    return (r5 << 11) | (g6 << 5) | b5


def expand_rgb565(c: int):
    """RGB565 → (r8, g8, b8) with bit replication."""
    r5 = (c >> 11) & 0x1F
    g6 = (c >>  5) & 0x3F
    b5 =  c        & 0x1F
    r = (r5 << 3) | (r5 >> 2)
    g = (g6 << 2) | (g6 >> 4)
    b = (b5 << 3) | (b5 >> 2)
    return r & 0xFF, g & 0xFF, b & 0xFF


def build_palette(e0: int, e1: int):
    """Build 4-colour palette using 3/8+5/8 weights (matching Wii GX decode)."""
    r0, g0, b0 = expand_rgb565(e0)
    r1, g1, b1 = expand_rgb565(e1)

    r2 = (5 * r0 + 3 * r1) >> 3
    g2 = (5 * g0 + 3 * g1) >> 3
    b2 = (5 * b0 + 3 * b1) >> 3

    r3 = (3 * r0 + 5 * r1) >> 3
    g3 = (3 * g0 + 5 * g1) >> 3
    b3 = (3 * b0 + 5 * b1) >> 3

    return [(r0,g0,b0), (r1,g1,b1), (r2,g2,b2), (r3,g3,b3)]


def encode_dxt1_block(pixels):
    """
    Encode a flat list of 16 (r,g,b) tuples (row-major 4x4) → bytes[8].
    Returns the 8-byte DXT1 block.
    """
    # 1. Bounding box
    rmin = gmin = bmin = 255
    rmax = gmax = bmax = 0
    for r, g, b in pixels:
        rmin, rmax = min(r, rmin), max(r, rmax)
        gmin, gmax = min(g, gmin), max(g, gmax)
        bmin, bmax = min(b, bmin), max(b, bmax)

    # 2. Quantise endpoints (round-to-nearest)
    e0 = pack_rgb565(rmax, gmax, bmax)
    e1 = pack_rgb565(rmin, gmin, bmin)

    # 3. Ensure 4-colour mode: e0 > e1
    if e0 < e1:
        e0, e1 = e1, e0
    if e0 == e1:
        if e0 < 0xFFFF:
            e0 += 1
        else:
            e1 -= 1

    # 4. Build palette
    palette = build_palette(e0, e1)

    # 5. Assign indices (min squared RGB distance)
    idx_bytes = []
    for row in range(4):
        byte = 0
        for col in range(4):
            r, g, b = pixels[row * 4 + col]
            best, best_err = 0, 10**9
            for k, (pr, pg, pb) in enumerate(palette):
                err = (r-pr)**2 + (g-pg)**2 + (b-pb)**2
                if err < best_err:
                    best_err, best = err, k
            byte |= best << (col * 2)
        idx_bytes.append(byte & 0xFF)

    # 6. Pack big-endian
    return bytes([e0 >> 8, e0 & 0xFF, e1 >> 8, e1 & 0xFF] + idx_bytes)


def cmpr_encode(rgb: bytes, width: int, height: int) -> bytes:
    """Encode width×height RGB888 image to GX_TF_CMPR."""
    assert width % 8 == 0 and height % 8 == 0
    out = bytearray()
    stx = width  // 8
    sty = height // 8

    for ty in range(sty):
        for tx in range(stx):
            for by in range(2):
                for bx in range(2):
                    px = tx * 8 + bx * 4
                    py = ty * 8 + by * 4
                    block = []
                    for y in range(4):
                        for x in range(4):
                            i = ((py + y) * width + (px + x)) * 3
                            block.append((rgb[i], rgb[i+1], rgb[i+2]))
                    out += encode_dxt1_block(block)
    return bytes(out)


def decode_dxt1_block(data: bytes, offset: int):
    """Decode one 8-byte DXT1 block → 16 (r,g,b) tuples."""
    e0 = (data[offset] << 8) | data[offset+1]
    e1 = (data[offset+2] << 8) | data[offset+3]

    if e0 > e1:
        palette = build_palette(e0, e1)
    else:
        r0,g0,b0 = expand_rgb565(e0)
        r1,g1,b1 = expand_rgb565(e1)
        palette = [
            (r0,g0,b0), (r1,g1,b1),
            ((r0+r1)//2, (g0+g1)//2, (b0+b1)//2),
            (0,0,0),
        ]

    pixels = []
    for row in range(4):
        ib = data[offset + 4 + row]
        for col in range(4):
            idx = (ib >> (col * 2)) & 3
            pixels.append(palette[idx])
    return pixels


def cmpr_decode(data: bytes, width: int, height: int) -> bytearray:
    """Decode GX_TF_CMPR → RGB888 bytearray."""
    out = bytearray(width * height * 3)
    stx = width  // 8
    sty = height // 8
    pos = 0

    for ty in range(sty):
        for tx in range(stx):
            for by in range(2):
                for bx in range(2):
                    pixels = decode_dxt1_block(data, pos)
                    pos += 8
                    px = tx * 8 + bx * 4
                    py = ty * 8 + by * 4
                    for row in range(4):
                        for col in range(4):
                            i = ((py + row) * width + (px + col)) * 3
                            out[i], out[i+1], out[i+2] = pixels[row*4+col]
    return out


# ---------------------------------------------------------------------------
# Test harness
# ---------------------------------------------------------------------------

s_pass = s_fail = 0

def check(name: str, cond: bool, detail: str = ""):
    global s_pass, s_fail
    tag = "PASS" if cond else "FAIL"
    line = f"  {tag}  {name}"
    if detail:
        line += f"  ({detail})"
    print(line)
    if cond:
        s_pass += 1
    else:
        s_fail += 1


def max_err(a: bytes, b: bytes) -> int:
    return max(abs(int(x) - int(y)) for x, y in zip(a, b))


def fill_solid(w, h, r, g, b) -> bytes:
    return bytes([r, g, b] * (w * h))


# ---------------------------------------------------------------------------
# Test 1 — solid colours
# ---------------------------------------------------------------------------

def test_solid_colors():
    print("\n[1] solid colours (round-trip error ≤ 8)")
    cases = [
        ("red",   255,   0,   0),
        ("green",   0, 255,   0),
        ("blue",    0,   0, 255),
        ("white", 255, 255, 255),
        ("black",   0,   0,   0),
        ("grey",  128, 128, 128),
        ("mid",   100,  50,  25),
    ]
    for name, r, g, b in cases:
        src  = fill_solid(8, 8, r, g, b)
        cmpr = cmpr_encode(src, 8, 8)
        dst  = cmpr_decode(cmpr, 8, 8)
        err  = max_err(src, dst)
        check(f"solid_{name}", err <= 8, f"err={err}")


# ---------------------------------------------------------------------------
# Test 2 — palette weights
# ---------------------------------------------------------------------------

def test_palette_weights():
    print("\n[2] 3/8+5/8 palette interpolation")

    # Verify the formula directly, no encoding involved
    # e0 > e1, both with known values
    e0 = pack_rgb565(200, 160, 80)
    e1 = pack_rgb565( 40,  20, 10)
    assert e0 > e1, "test assumption: e0>e1"

    r0,g0,b0 = expand_rgb565(e0)
    r1,g1,b1 = expand_rgb565(e1)

    exp_c2 = ((5*r0+3*r1)>>3, (5*g0+3*g1)>>3, (5*b0+3*b1)>>3)
    exp_c3 = ((3*r0+5*r1)>>3, (3*g0+5*g1)>>3, (3*b0+5*b1)>>3)

    palette = build_palette(e0, e1)

    check("c0 == expand(e0)", palette[0] == (r0,g0,b0))
    check("c1 == expand(e1)", palette[1] == (r1,g1,b1))
    check("c2 == (5*c0+3*c1)>>3", palette[2] == exp_c2,
          f"got={palette[2]} want={exp_c2}")
    check("c3 == (3*c0+5*c1)>>3", palette[3] == exp_c3,
          f"got={palette[3]} want={exp_c3}")

    # Also verify these differ from standard DXT1 (2/3+1/3) weights
    std_c2 = ((2*r0+r1+1)//3, (2*g0+g1+1)//3, (2*b0+b1+1)//3)
    std_c3 = ((r0+2*r1+1)//3, (g0+2*g1+1)//3, (b0+2*b1+1)//3)
    print(f"      5/8+3/8 c2={exp_c2}  standard 2/3+1/3 c2={std_c2}")
    print(f"      3/8+5/8 c3={exp_c3}  standard 1/3+2/3 c3={std_c3}")


# ---------------------------------------------------------------------------
# Test 3 — round-to-nearest pack_rgb565
# ---------------------------------------------------------------------------

def test_rtn_packing():
    print("\n[3] round-to-nearest RGB565 packing")
    errors = []
    for r in range(256):
        for b in range(256):
            packed = pack_rgb565(r, 0, b)
            re, _, be = expand_rgb565(packed)
            err_r = abs(r - re)
            err_b = abs(b - be)
            # Round-trip error ≤ half an LSB of the 5-bit channel (~4)
            if err_r > 4:
                errors.append(f"r={r} err={err_r}")
            if err_b > 4:
                errors.append(f"b={b} err={err_b}")
    check("5-bit channel round-trip error ≤ 4", not errors,
          f"{len(errors)} violations" if errors else "")

    errors6 = []
    for g in range(256):
        packed = pack_rgb565(0, g, 0)
        _, ge, _ = expand_rgb565(packed)
        err = abs(g - ge)
        if err > 2:
            errors6.append(f"g={g} err={err}")
    check("6-bit channel round-trip error ≤ 2", not errors6,
          f"{len(errors6)} violations" if errors6 else "")


# ---------------------------------------------------------------------------
# Test 4 — gradient
# ---------------------------------------------------------------------------

def test_gradient():
    # DXT1 bbox-encoder worst-case: 18.75% of the block's colour range per channel.
    # With 3/8+5/8 palette the four colours sit at 0%, 37.5%, 62.5%, 100%.
    # In a multi-channel gradient 3D nearest-neighbour can give up ~35 in one
    # channel to minimise overall 3D error.  Limit 40 covers this with margin.
    print("\n[4] horizontal gradient 16x8 (max error <= 40)")
    src = bytearray()
    for y in range(8):
        for x in range(16):
            v = x * 255 // 15
            src += bytes([v, v // 2, 255 - v])
    cmpr = cmpr_encode(bytes(src), 16, 8)
    dst  = cmpr_decode(cmpr, 16, 8)
    err  = max_err(src, dst)
    check(f"gradient_max_err <= 40", err <= 40, f"err={err}")


# ---------------------------------------------------------------------------
# Test 5 — output size
# ---------------------------------------------------------------------------

def test_output_size():
    print("\n[5] output byte count = width×height/2")
    for w, h in [(8,8), (16,8), (16,16), (32,32)]:
        src  = fill_solid(w, h, 128, 64, 32)
        cmpr = cmpr_encode(src, w, h)
        want = w * h // 2
        check(f"size_{w}x{h}", len(cmpr) == want, f"got={len(cmpr)} want={want}")


# ---------------------------------------------------------------------------
# Test 6 — super-tile order
# ---------------------------------------------------------------------------

def test_super_tile_order():
    print("\n[6] super-tile order: red|blue side-by-side 16×8")
    src = bytearray()
    for y in range(8):
        for x in range(8):
            src += bytes([255, 0, 0])   # left tile: red
        for x in range(8):
            src += bytes([0, 0, 255])   # right tile: blue
    cmpr = cmpr_encode(bytes(src), 16, 8)
    dst  = cmpr_decode(cmpr, 16, 8)

    left_ok = right_ok = True
    for y in range(8):
        for x in range(8):
            r,g,b = dst[(y*16+x)*3], dst[(y*16+x)*3+1], dst[(y*16+x)*3+2]
            if r < 240 or g > 8 or b > 8:
                left_ok = False
        for x in range(8, 16):
            r,g,b = dst[(y*16+x)*3], dst[(y*16+x)*3+1], dst[(y*16+x)*3+2]
            if r > 8 or g > 8 or b < 240:
                right_ok = False

    check("left_tile_is_red",   left_ok)
    check("right_tile_is_blue", right_ok)


# ---------------------------------------------------------------------------
# Test 7 — e0 > e1 invariant for non-transparent blocks
# ---------------------------------------------------------------------------

def test_endpoints_ordered():
    print("\n[7] e0>e1 invariant (4-colour mode) for all solid blocks")
    failures = 0
    for r in range(0, 256, 32):
        for g in range(0, 256, 32):
            for b in range(0, 256, 32):
                src  = fill_solid(8, 8, r, g, b)
                cmpr = cmpr_encode(src, 8, 8)
                # Check the first block (bytes 0-1 vs 2-3)
                e0 = (cmpr[0] << 8) | cmpr[1]
                e1 = (cmpr[2] << 8) | cmpr[3]
                if e0 <= e1:
                    failures += 1
    check("all_solid_blocks_use_4colour_mode", failures == 0,
          f"{failures} blocks had e0≤e1")


# ---------------------------------------------------------------------------
# main
# ---------------------------------------------------------------------------

def main():
    print("=== CMPR smoke test (Python independent implementation) ===")
    test_solid_colors()
    test_palette_weights()
    test_rtn_packing()
    test_gradient()
    test_output_size()
    test_super_tile_order()
    test_endpoints_ordered()
    print(f"\n=== Results: {s_pass} passed, {s_fail} failed ===")
    sys.exit(1 if s_fail else 0)


if __name__ == "__main__":
    main()

File: tools/cmpr_smoke_test.c
/*
 * cmpr_smoke_test.c — host-side verification of the CMPR encoder
 *
 * Compile on Linux / macOS:
 *   gcc -std=c11 -Wall -Wextra -o cmpr_smoke_test \
 *       cmpr_smoke_test.c ../code/renderer/wii_cmpr.c
 *
 * Compile inside devkitPro's MSYS2 (Cygwin-based GCC):
 *   GCC_INC=/c/devkitPro/msys2/usr/lib/gcc/x86_64-pc-cygwin/15.2.0/include
 *   SYS_INC=/c/devkitPro/msys2/usr/include
 *   gcc -std=c11 -Wall -isystem "$GCC_INC" -isystem "$SYS_INC" \
 *       -o cmpr_smoke_test.exe \
 *       cmpr_smoke_test.c ../code/renderer/wii_cmpr.c
 *
 * Run:
 *   ./cmpr_smoke_test  (or ./cmpr_smoke_test.exe on Windows)
 *
 * Tests
 * ─────
 *  1. solid_colors     — solid R, G, B, white, black, grey
 *                        quantisation round-trip should be ≤ 8 per channel
 *  2. two_color_block  — known endpoints; verify 3/8+5/8 palette values exactly
 *  3. palette_weights  — explicit check: c2 = (5*c0+3*c1)>>3, c3 = (3*c0+5*c1)>>3
 *  4. gradient_8x8     — horizontal ramp; max per-channel error ≤ 16
 *  5. encode_size      — output byte count matches width*height/2
 *  6. super_tile_order — two solid-colour 8×8 tiles side-by-side;
 *                        verify each super-tile decodes to its own colour
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

/* Include the encoder/decoder — no Wii headers needed */
#include "../code/renderer/wii_cmpr.h"

/* =========================================================================
   Helpers
   ========================================================================= */

static int s_pass = 0, s_fail = 0;

static void report(const char *name, int cond)
{
    if (cond) {
        printf("  PASS  %s\n", name);
        s_pass++;
    } else {
        printf("  FAIL  %s\n", name);
        s_fail++;
    }
}

/* Maximum per-channel absolute error across the entire image */
static int max_error(const uint8_t *a, const uint8_t *b, int n_pixels)
{
    int mx = 0;
    for (int i = 0; i < n_pixels * 3; i++) {
        int e = (int)a[i] - (int)b[i];
        if (e < 0) e = -e;
        if (e > mx) mx = e;
    }
    return mx;
}

/* Fill an 8×8 RGB block with a solid colour */
static void fill_solid(uint8_t *buf, uint8_t r, uint8_t g, uint8_t b)
{
    for (int i = 0; i < 8 * 8; i++) {
        buf[i*3+0] = r;
        buf[i*3+1] = g;
        buf[i*3+2] = b;
    }
}

/* =========================================================================
   Test 1 — solid colours: round-trip error ≤ 8 per channel
   ========================================================================= */
static void test_solid_colors(void)
{
    printf("\n[1] solid colours\n");

    static const struct { const char *name; uint8_t r,g,b; } cases[] = {
        { "red",   255,   0,   0 },
        { "green",   0, 255,   0 },
        { "blue",    0,   0, 255 },
        { "white", 255, 255, 255 },
        { "black",   0,   0,   0 },
        { "grey",  128, 128, 128 },
        { "mid-r", 100,  50,  25 },
        { NULL,      0,   0,   0 }
    };

    uint8_t src[8*8*3], cmpr[8*8/2], dst[8*8*3];

    for (int i = 0; cases[i].name; i++) {
        fill_solid(src, cases[i].r, cases[i].g, cases[i].b);
        CMPR_Encode(src, 8, 8, cmpr);
        CMPR_Decode(cmpr, 8, 8, dst);

        int err = max_error(src, dst, 8*8);
        char label[64];
        snprintf(label, sizeof(label), "solid_%s (err=%d)", cases[i].name, err);
        report(label, err <= 8);
    }
}

/* =========================================================================
   Test 2 — palette weights: explicit bit-exact check
   ========================================================================= */

/*
 * Verify that for a block whose two pixels span the full endpoint range,
 * the decoded intermediate palette entries match (5*c0+3*c1)>>3 and
 * (3*c0+5*c1)>>3 exactly.
 *
 * We make the block degenerate:
 *   - top-left  (row0,col0) = endpoint max
 *   - bottom-right (row3,col3) = endpoint min
 *   - all other pixels = those two colours alternately so the bbox is stable
 * Then we decode the block and read back c2 and c3 from the pixels that
 * were assigned index 2 and index 3 by the encoder.
 *
 * Rather than inspecting the raw CMPR bytes, we verify the decoded palette
 * by building the palette ourselves and comparing.
 */
static void test_palette_weights(void)
{
    printf("\n[2] 3/8+5/8 palette weights\n");

    /*
     * Choose two colours that are far apart so the encoder must use
     * both intermediate palette entries.
     *
     * c0_rgb = (200, 160, 80)   → pack/expand → some value we can inspect
     * c1_rgb = ( 40,  20, 10)
     *
     * Fill the 8×8 block with a 2×2 checkerboard so both endpoints are
     * well-represented and the bbox is tight.
     */
    const uint8_t C0[3] = {200, 160, 80};
    const uint8_t C1[3] = { 40,  20, 10};

    uint8_t src[8*8*3];
    for (int y = 0; y < 8; y++)
        for (int x = 0; x < 8; x++) {
            const uint8_t *c = ((x ^ y) & 1) ? C1 : C0;
            src[(y*8+x)*3+0] = c[0];
            src[(y*8+x)*3+1] = c[1];
            src[(y*8+x)*3+2] = c[2];
        }

    uint8_t cmpr[8*8/2];
    CMPR_Encode(src, 8, 8, cmpr);

    /*
     * Extract the first block's endpoints (bytes 0-3 of the CMPR buffer).
     * They are big-endian RGB565.
     */
    uint16_t e0 = ((uint16_t)cmpr[0] << 8) | cmpr[1];
    uint16_t e1 = ((uint16_t)cmpr[2] << 8) | cmpr[3];

    /* Expand */
    uint8_t r0,g0,b0, r1,g1,b1;
    {
        uint8_t r5 = (e0>>11)&0x1F, g6=(e0>>5)&0x3F, b5=e0&0x1F;
        r0=(r5<<3)|(r5>>2); g0=(g6<<2)|(g6>>4); b0=(b5<<3)|(b5>>2);
    }
    {
        uint8_t r5 = (e1>>11)&0x1F, g6=(e1>>5)&0x3F, b5=e1&0x1F;
        r1=(r5<<3)|(r5>>2); g1=(g6<<2)|(g6>>4); b1=(b5<<3)|(b5>>2);
    }

    /* Expected c2 and c3 */
    uint8_t exp_r2 = (uint8_t)((5u*r0 + 3u*r1) >> 3);
    uint8_t exp_g2 = (uint8_t)((5u*g0 + 3u*g1) >> 3);
    uint8_t exp_b2 = (uint8_t)((5u*b0 + 3u*b1) >> 3);
    uint8_t exp_r3 = (uint8_t)((3u*r0 + 5u*r1) >> 3);
    uint8_t exp_g3 = (uint8_t)((3u*g0 + 5u*g1) >> 3);
    uint8_t exp_b3 = (uint8_t)((3u*b0 + 5u*b1) >> 3);

    /* Decode and find pixels assigned to indices 2 and 3 */
    uint8_t dst[8*8*3];
    CMPR_Decode(cmpr, 8, 8, dst);

    /* The decoder uses the same formula — so any pixel with idx=2 will have
     * colour (exp_r2, exp_g2, exp_b2).  We find such pixels via the raw
     * index byte of block 0 (cmpr byte 4). */
    int found_idx2 = 0, found_idx3 = 0;
    int ok2 = 1, ok3 = 1;

    for (int y = 0; y < 4; y++) {
        uint8_t ib = cmpr[4 + y];
        for (int x = 0; x < 4; x++) {
            int idx = (ib >> (x*2)) & 3;
            const uint8_t *dp = dst + (y*8+x)*3;
            if (idx == 2) {
                found_idx2 = 1;
                if (dp[0]!=exp_r2 || dp[1]!=exp_g2 || dp[2]!=exp_b2) ok2=0;
            }
            if (idx == 3) {
                found_idx3 = 1;
                if (dp[0]!=exp_r3 || dp[1]!=exp_g3 || dp[2]!=exp_b3) ok3=0;
            }
        }
    }

    report("endpoints_are_ordered (e0>e1)", e0 > e1);
    report("c2_pixel_matches_5/8+3/8 formula (if used)", !found_idx2 || ok2);
    report("c3_pixel_matches_3/8+5/8 formula (if used)", !found_idx3 || ok3);

    printf("      e0=0x%04X (%u,%u,%u)  e1=0x%04X (%u,%u,%u)\n",
           e0, r0,g0,b0, e1, r1,g1,b1);
    printf("      expected c2=(%u,%u,%u)  c3=(%u,%u,%u)\n",
           exp_r2,exp_g2,exp_b2, exp_r3,exp_g3,exp_b3);
}

/* =========================================================================
   Test 3 — gradient: max error ≤ 16
   ========================================================================= */
static void test_gradient(void)
{
    printf("\n[3] horizontal gradient (16×8)\n");

    uint8_t src[16*8*3];
    for (int y = 0; y < 8; y++)
        for (int x = 0; x < 16; x++) {
            uint8_t v = (uint8_t)(x * 255 / 15);
            src[(y*16+x)*3+0] = v;
            src[(y*16+x)*3+1] = v / 2;
            src[(y*16+x)*3+2] = (uint8_t)(255 - v);
        }

    uint8_t cmpr[16*8/2];
    CMPR_Encode(src, 16, 8, cmpr);

    uint8_t dst[16*8*3];
    CMPR_Decode(cmpr, 16, 8, dst);

    int err = max_error(src, dst, 16*8);
    char label[64];
    snprintf(label, sizeof(label), "gradient max_err=%d (limit 16)", err);
    report(label, err <= 16);
}

/* =========================================================================
   Test 4 — output size
   ========================================================================= */
static void test_encode_size(void)
{
    printf("\n[4] output size\n");

    /* We verify the caller-contract: output must be width*height/2 bytes.
     * Write a sentinel past the expected end and confirm the encoder
     * doesn't clobber it. */
    const int W = 16, H = 16;
    size_t expected = (size_t)(W * H / 2);
    size_t buf_size = expected + 16;

    uint8_t *src  = (uint8_t *)calloc(W * H * 3, 1);
    uint8_t *cmpr = (uint8_t *)calloc(buf_size, 1);
    if (!src || !cmpr) { report("alloc", 0); free(src); free(cmpr); return; }

    /* Fill sentinel */
    memset(cmpr + expected, 0xDE, 16);

    CMPR_Encode(src, W, H, cmpr);

    int sentinel_ok = 1;
    for (int i = 0; i < 16; i++)
        if (cmpr[expected + i] != 0xDE) { sentinel_ok = 0; break; }

    char label[64];
    snprintf(label, sizeof(label),
             "output_size %dx%d → %zu bytes, sentinel intact",
             W, H, expected);
    report(label, sentinel_ok);

    free(src);
    free(cmpr);
}

/* =========================================================================
   Test 5 — super-tile order: two side-by-side 8×8 tiles
   ========================================================================= */
static void test_super_tile_order(void)
{
    printf("\n[5] super-tile order (16×8 with two distinct tiles)\n");

    /* Left tile: solid red (255,0,0), right tile: solid blue (0,0,255) */
    uint8_t src[16*8*3];
    for (int y = 0; y < 8; y++) {
        for (int x = 0; x < 8; x++) {
            src[(y*16+x)*3+0] = 255; src[(y*16+x)*3+1] = 0; src[(y*16+x)*3+2] = 0;
        }
        for (int x = 8; x < 16; x++) {
            src[(y*16+x)*3+0] = 0; src[(y*16+x)*3+1] = 0; src[(y*16+x)*3+2] = 255;
        }
    }

    uint8_t cmpr[16*8/2];
    CMPR_Encode(src, 16, 8, cmpr);

    uint8_t dst[16*8*3];
    CMPR_Decode(cmpr, 16, 8, dst);

    /* Left region should decode to red, right to blue (within quant error) */
    int left_ok = 1, right_ok = 1;
    for (int y = 0; y < 8 && (left_ok || right_ok); y++) {
        for (int x = 0; x < 8; x++) {
            const uint8_t *p = dst + (y*16+x)*3;
            if (p[0] < 240 || p[1] > 8 || p[2] > 8) left_ok = 0;
        }
        for (int x = 8; x < 16; x++) {
            const uint8_t *p = dst + (y*16+x)*3;
            if (p[0] > 8 || p[1] > 8 || p[2] < 240) right_ok = 0;
        }
    }

    report("left_tile_is_red",  left_ok);
    report("right_tile_is_blue", right_ok);
}

/* =========================================================================
   Entry point
   ========================================================================= */

static int run_tests(void)
{
    printf("=== CMPR smoke test ===\n");

    test_solid_colors();
    test_palette_weights();
    test_gradient();
    test_encode_size();
    test_super_tile_order();

    printf("\n=== Results: %d passed, %d failed ===\n", s_pass, s_fail);
    return s_fail ? 1 : 0;
}

/*
 * MSYS2/Cygwin GCC links libcmain.o which provides the PE 'main' entry
 * and expects the user to supply WinMain.  Everywhere else, supply main.
 */
#if defined(__MSYS__) || defined(__CYGWIN__)
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
int WINAPI WinMain(HINSTANCE h, HINSTANCE p, LPSTR cmd, int show)
{
    (void)h; (void)p; (void)cmd; (void)show;
    return run_tests();
}
#else
int main(void) { return run_tests(); }
#endif
